boa/src/syntax/lexer/tests.rs

```diff
@@ -264,283 +152,387 @@ fn check_keywords() {
+
+    let expected = [
+        TokenKind::Keyword(Keyword::Let),
+        TokenKind::identifier("a"),
+        TokenKind::Punctuator(Punctuator::Assign),
+        TokenKind::string_literal("hello"),
+        TokenKind::Punctuator(Punctuator::Semicolon),
+    ];
+
+    expect_tokens(&mut lexer, &expected);
 }
 
 #[test]
 fn check_positions() {
     let s = r#"console.log("hello world\u{2764}"); // Test"#;
     // --------123456789
-    let mut lexer = Lexer::new(s);
-    lexer.lex().expect("failed to lex");
+    let mut lexer = Lexer::new(s.as_bytes());
+
     // The first column is 1 (not zero indexed)
-    assert_eq!(lexer.tokens[0].span(), span((1, 1), (1, 7)));
+    assert_eq!(lexer.next().unwrap().unwrap().span(), span((1, 1), (1, 8)));
 
     // Dot Token starts on column 8
-    assert_eq!(lexer.tokens[1].span(), span((1, 8), (1, 8)));
+    assert_eq!(lexer.next().unwrap().unwrap().span(), span((1, 8), (1, 9)));
 
     // Log Token starts on column 9
-    assert_eq!(lexer.tokens[2].span(), span((1, 9), (1, 11)));
+    assert_eq!(lexer.next().unwrap().unwrap().span(), span((1, 9), (1, 12)));
 
     // Open parenthesis token starts on column 12
-    assert_eq!(lexer.tokens[3].span(), span((1, 12), (1, 12)));
+    assert_eq!(
+        lexer.next().unwrap().unwrap().span(),
+        span((1, 12), (1, 13))
+    );
 
     // String token starts on column 13
-    assert_eq!(lexer.tokens[4].span(), span((1, 13), (1, 33)));
+    assert_eq!(
+        lexer.next().unwrap().unwrap().span(),
+        span((1, 13), (1, 34))
+    );
 
     // Close parenthesis token starts on column 34
-    assert_eq!(lexer.tokens[5].span(), span((1, 34), (1, 34)));
+    assert_eq!(
+        lexer.next().unwrap().unwrap().span(),
+        span((1, 34), (1, 35))
+    );
 
     // Semi Colon token starts on column 35
-    assert_eq!(lexer.tokens[6].span(), span((1, 35), (1, 35)));
+    assert_eq!(
+        lexer.next().unwrap().unwrap().span(),
+        span((1, 35), (1, 36))
+    );
 }
 
 #[test]
 #[ignore]
 fn two_divisions_in_expression() {
     let s = "    return a !== 0 || 1 / a === 1 / b;";
-    let mut lexer = Lexer::new(s);
-    lexer.lex().expect("failed to lex");
-    // dbg!(&lexer.tokens);
+    let mut lexer = Lexer::new(s.as_bytes());
 
-    assert_eq!(lexer.tokens[11].span(), span((1, 37), (1, 37)));
+    assert_eq!(
+        lexer.nth(11).unwrap().unwrap().span(),
+        span((1, 37), (1, 37))
+    );
 }
 
 #[test]
 fn check_line_numbers() {
     let s = "x\ny\n";
 
-    let mut lexer = Lexer::new(s);
-    lexer.lex().expect("failed to lex");
+    let mut lexer = Lexer::new(s.as_bytes());
 
-    assert_eq!(lexer.tokens[0].span(), span((1, 1), (1, 1)));
-    assert_eq!(lexer.tokens[1].span(), span((1, 2), (2, 1)));
-    assert_eq!(lexer.tokens[2].span(), span((2, 1), (2, 1)));
-    assert_eq!(lexer.tokens[3].span(), span((2, 2), (3, 1)));
+    assert_eq!(lexer.next().unwrap().unwrap().span(), span((1, 1), (1, 1)));
```

Comment:
Here it seems the old lexer used inclusive end bounds (issue #456 ) and that the new line symbols were returned as tokens? 

Am I right is saying that we should be using exclusive bounds and that control characters such as new line shouldn't be produced by lexer.next()?  (And therefore that this test is wrong and needs updating)

Reply:
We should use exclusive bounds, but line terminators must be produced by `lexer.next()`, since even if in most cases they are just white space, they have some roles in automatic semicolon insertion.

Reply:
Ah right ok, I think I'll need to add some tests in for this to check line terminators are returned (in most tests currently they aren't).